dynamic_array.c

```diff
@@ -0,0 +1,340 @@
+#include "dynamic_array.h"
+
+// NOTE: When we destroy items, we call the destructor for them 
+// iff the destructor is not NULL.
+
+typedef void (*destr)(void *);
```

Comment:
Here, you have defined the type explicitly in the initialiser signature (`void (*destructor)(void *)`) then declared a typedef privately within the C file.  It would be better to define the typedef in the header (e.g. `typedef void buffer_item_destructor(void *);`) then use that type in parameter lists and in the struct definition (`buffer_item_destructor *destructor`).

Since destructors are very common and go beyond the scope of a container library, I've added `typedef void destructor(void *);` to the `cstd` module (commit 25499aa), so we have a standard destructor signature/type for all modules to use.